File: Leetcode/dfs_recursion_backtracking/690_Employee Importance/Solution.java
/*
// Employee info
class Employee {
    // It's the unique id of each node;
    // unique id of this employee
    public int id;
    // the importance value of this employee
    public int importance;
    // the id of direct subordinates
    public List<Integer> subordinates;
};
*/
class Solution {
    public int getImportance(List<Employee> employees, int id) {
        HashMap<Integer, Employee> ids = new HashMap<>();
        for(Employee emp: employees){
            ids.put(emp.id, emp);
        }
        int res = dfs(ids, id);
        return res;
    }
        private int dfs(HashMap<Integer, Employee> ids, int id){
            Employee cur = ids.get(id);
            int res = cur.importance;
            if(cur.subordinates.size()==0)
                return res;
            for(int ID: cur.subordinates){
                res += dfs(ids, ID);
            }
            return res;
        }
}

File: Leetcode/dfs_recursion_backtracking/690_Employee Importance/Solution.py
"""
# Employee info
class Employee(object):
    def __init__(self, id, importance, subordinates):
        # It's the unique id of each node.
        # unique id of this employee
        self.id = id
        # the importance value of this employee
        self.importance = importance
        # the id of direct subordinates
        self.subordinates = subordinates
"""
class Solution(object):
    def getImportance(self, employees, id):
        """
        :type employees: Employee
        :type id: int
        :rtype: int
        """
        ids = {}
        for employee in employees:
            ids[employee.id] = employee
        def dfs(cur_id):
            cur = ids[cur_id]
            res = cur.importance
            if not cur.subordinates:
                return res
            for ID in cur.subordinates:
                res += dfs(ID)
            return res
        return dfs(id)
